duckduckgo/Android
# Update tracker detection to use new tds data set

**Task/Issue URL:** 

- New TDS Support: https://app.asana.com/0/361428290920652/1147045768949348
- Embed TDS in App: https://app.asana.com/0/361428290920652/1147872489592776
- Add global/temp whitelist: https://app.asana.com/0/361428290920652/1147872489592777 


**Tech Design URL:**

- https://app.asana.com/0/361428290920652/1132331070554722

**Description**:

- Implement new tracker data set and entities
- Embed TDS data in app to protect first run users
- Remove legacy tracker detection code and libraries
- Add Global whitelist support to allow us to publish temporary fixes and recover quickly from any issues in the TDS data set.

**Steps to test this PR**

**Test Setup:**
- in `PrivacyDashboardViewModel` set `private const val LEADERBOARD_MIN_DOMAINS_EXCLUSIVE = 3` 

**Test Database Migration:**
- Install the app from develop
- Run the app and load some sites (at least 3 including one with many trackers)
- Now install the latest version, run the app and ensure everything runs fine
- Check that the network leaderboard pills in the Privacy Dashboard are reset
- Run the app and load some sites (at least 3 including one with many trackers)
- Check that the network leaderboard pills are now populated

**Test that trackers are detected and displayed as expected:**
- Open a site with many trackers such as cnn.com
- Tap on the privacy grade to open the Privacy Dashboard
- Ensure that trackers are listed
- Ensure that the network list shows major trackers first and that icons (where they exist) are populated

**Test that embedded tds still works even if download fails:**
- Use charles to block calls to `https://staticcdn.duckduckgo.com/trackerblocking/v2/tds.json`
- Install the app and ensure that tracker blocking still works as normal 

**Test that tds is updated from backend:**
- Update the embedded tds.json file content to be empty json`{}`
- Run a clean install of the app and allow app config download to complete
- Visit a site with trackers and ensure that these are still detected

**Test temporary/global whitelist:**
- Visit a site in our temporary whitelist https://duckduckgo.com//contentblocking/trackers-whitelist-temporary.txt e.g fidelity.com and ensure that no trackers are blocked on it
- Now cleanly install the app and block the call to the temporary whitelist
- Wait for the rest of the app config to download
- Visit the site previously tested and check that trackers are now blocked on it

---
###### Internal references:
[Software Engineering Expectations](https://app.asana.com/0/59792373528535/199064865822552)
[Technical Design Template](https://app.asana.com/0/59792373528535/184709971311943)
